Assets/MRTK/Core/Providers/InputAnimation/InputAnimation.cs

```diff
@@ -34,25 +34,123 @@ public class InputAnimationMarker
+            public void Optimize(float positionThreshold, float rotationThreshold, int partitionSize)
+            {
+                OptimizePositionCurve(ref PositionX, ref PositionY, ref PositionZ, positionThreshold, partitionSize);
+                OptimizeRotationCurve(ref RotationX, ref RotationY, ref RotationZ, ref RotationW, rotationThreshold, partitionSize);
+            }
+
+            public MixedRealityPose Evaluate(float time)
+            {
+                float px = PositionX.Evaluate(time);
+                float py = PositionY.Evaluate(time);
+                float pz = PositionZ.Evaluate(time);
+                float rx = RotationX.Evaluate(time);
+                float ry = RotationY.Evaluate(time);
+                float rz = RotationZ.Evaluate(time);
+                float rw = RotationW.Evaluate(time);
+
+                var pose = new MixedRealityPose();
+                
+                pose.Position = new Vector3(px, py, pz);
+                pose.Rotation = new Quaternion(rx, ry, rz, rw);
+                pose.Rotation.Normalize();
+                
+                return pose;
+            }
+        }
+        
+        public class RayCurves
+        {
+            public AnimationCurve OriginX = new AnimationCurve();
+            public AnimationCurve OriginY = new AnimationCurve();
+            public AnimationCurve OriginZ = new AnimationCurve();
+            public AnimationCurve DirectionX = new AnimationCurve();
+            public AnimationCurve DirectionY = new AnimationCurve();
+            public AnimationCurve DirectionZ = new AnimationCurve();
+
+            public void AddKey(float time, Ray ray)
+            {
+                AddVectorKey(OriginX, OriginY, OriginZ, time, ray.origin);
+                AddVectorKey(DirectionX, DirectionY, DirectionZ, time, ray.direction);
+            }
+            
+            /// <summary>
+            /// Optimizes the set of curves.
+            /// </summary>
+            /// <param name="originThreshold">The maximum permitted error between the origins of the old and new curves, in units.</param>
+            /// <param name="directionThreshold">The maximum permitted error between the directions of the old and new curves, in degrees.</param>
+            /// <param name="partitionSize">The size of the partitions of the curves that will be optimized independently. Larger values will optimize the curves better, but may take longer.</param>
+            public void Optimize(float originThreshold, float directionThreshold, int partitionSize)
+            {
+                OptimizePositionCurve(ref OriginX, ref OriginY, ref OriginZ, originThreshold, partitionSize);
+                OptimizeDirectionCurve(ref DirectionX, ref DirectionY, ref DirectionZ, directionThreshold, partitionSize);
+            }
+
+            public Ray Evaluate(float time)
```

Comment:
Might be good to do another pass to find all public methods classes and properties that are missing comments, though I think the CI server should be catching all these for you.